### PR TITLE
Optimize pytorch layer_norm forward

### PR DESCRIPTION
**Pull request by xiaomengy**
Summary:
Seperate from D15194600
Optimize pytorch layer_norm op part 1:
  optimize layer_norm_forward_cpu
  import Eigen Maps for the performance of reduction

Differential Revision: D15290608

### CODE REVIEW

**Review by soumith — CHANGES_REQUESTED** [review]
this PR now introduces Eigen into pytorch optimizations -- without discussion.
The previous comments of moving to native/cpu and using dispatcher are not addressed

**Comment by BIT-silence** [comment]
Removed Eigen now. For reduce case, I found that even using Vec256 is still slower than Eigen. So maybe we can consider to discuss or figure out how to make it better.

File: aten/src/ATen/native/layer_norm.cpp [hunk 1]
```diff
@@ -0,0 +1,148 @@
+#include <ATen/NativeFunctions.h>
```
Comment (VitalyFedyunin) [inline aten/src/ATen/native/layer_norm.cpp]: Please move LayerNormForwardCPUImpl and all CPU logic into the /cpu subfolder using DECLARE_DISPATCH logic (see CopyKernel for the reference). It will allow to utilize AVX2 instructions and other optimizations of OSS build.

File: aten/src/ATen/native/layer_norm.cpp [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+      X_dims.cbegin(),
+      X_dims.cbegin() + axis,
+      1LL,
+      std::multiplies<std::int64_t>());
+  const std::int64_t N = std::accumulate(
+      X_dims.cbegin() + axis,
+      X_dims.cend(),
+      1LL,
+      std::multiplies<std::int64_t>());
+  Tensor Y = at::empty_like(X);
```
Comment (VitalyFedyunin) [inline aten/src/ATen/native/layer_norm.cpp]: I'm quite sure you want to call at:native::empty_like here and avoid all tracing/profiling checks.

**Review by VitalyFedyunin — CHANGES_REQUESTED** [review]
This PR requires new tests to be written for native_layer_norm

**Comment by cpuhrsch** [comment]
@BIT-silence - Is this a copy of the C2 kernels?

**Comment by VitalyFedyunin** [comment]
You are exposing native_layer_norm function to the user level api which seems to be requiring contiguous inputs. Please add tests to cover APi calls as well as contiguous checks. Alternatively if intend is to introduce fast kernel, implement all code as kernel and avoid creating a new native function.

Also it would be nice to see benchmarks that compares calling old and new layer_norm code.

**Comment by BIT-silence** [comment]
> @BIT-silence - Is this a copy of the C2 kernels?

Logically it is, while C2 kernel is implemented by using Eigen lib. I tested the performance difference, for the rowwise moments part, Eigen version will be faster than this version by using compiler's auto vectorization, elementwise affine part performs the same. While I also tried to use Vec256 in rowwise moments part, it actually a little slower than the for-loop with auto vectorization.

File: aten/src/ATen/native/layer_norm.cpp (hunk 1 above)
Comment (BIT-silence) [reply aten/src/ATen/native/layer_norm.cpp]: Done

**Comment by BIT-silence** [comment]
Thanks for the advice. I have removed native_layer_norm function. Later I will add the backward part for layer_norm then do autograd for layer_norm instead of batch_norm.

Some benchmark result for this change.
input shape = [64, 128, 56, 56] and normalized_shape = [128, 56, 56] with elementwise_affine=True,
on devvm forward time from 350ms to 87.6ms. And with this approach, it can let hugging face BERT model forward about 12% faster.

**Review by soumith — APPROVED** [review]
thanks for making the changes. From my side this is good to go.

**Comment by BIT-silence** [comment]
> This PR requires new tests to be written for native_layer_norm

Actually I have removed native_layer_norm. Currently we don't expose new functions and add a fast path for layer_norm. Is that fine?

**Comment by fmassa** [comment]
We should ideally test that the fast path gives the same results as the slow path. But maybe this is implicitly tested in the cuda tests

**Comment by BIT-silence** [comment]
> We should ideally test that the fast path gives the same results as the slow path. But maybe this is implicitly tested in the cuda tests

Actually not only cuda, but also jit test covers that.

**Comment by BIT-silence** [comment]
Since I will add the grad part for this fast path and then all the layer_norm on CPU should go through this path. As the current fast path is actually tested by existing tests, I'm wondering if there is any need to add some specific test for that.
